### PR TITLE
Converts Core and Web projects to use project.json, fixes MSBuild, more c# 6 support

### PR DESCRIPTION
**Pull request by Shazwazza**
* Changes Umbraco.Core to use project.json for package references, instead of nuget (next is the Web proj)
* Changes the build to use correct version of MSBuild
* Updates the Web.UI project to reference Microsoft.CodeDom.Providers.DotNetCompilerPlatform so that we can use c# 6 in views
* Fixes MSBuild script to not use the old tool version which forces the non recent msbuild version
* Updates all other projects to use auto-package restore instead of the legacy nuget target restore
* Ensures project.lock.json is ignore

works on my machine - just need to test a clean check and run Build.bat to make sure it works - it will so long as msbuild generates the project.lock.json files - which it should.

this gets us a small step further to forking v8 into a v9 fork to support cross compilation to dotnetcore

### CODE REVIEW

**Comment by readingdancer** [comment]
Hi @Shazwazza, interesting commit :)

When this finally gets released it will be important to let everyone know they will need to upgrade to NuGet 3.0 this is no doubt fine for development machines, but maybe more work if NuGet is part of an automated build process and it could be an area people forget they will need to update.

In case anyone else looks at this commit, here's a link to the [Project.json info on the NuGet site](http://docs.nuget.org/consume/ProjectJson-Intro).

**Comment by Shazwazza** [comment]
yes well in v8 we are using csharp 6, VS 2015 is a requirement, etc.., etc...  the build script should always 'just work' and we'll make sure that is the case.

**Comment by readingdancer** [comment]
Hi @Shazwazza I was not meaning the build script for building Umbraco, I was meaning end project builds that include Umbraco. We have a lot of projects that we support and currently upgrade using NuGet which works nicely, when changes are committed to GIT the build server will use NuGet to get the updated packages, if U8 is using the new format then all the projects will need to be updated to use the new NuGet format if the old format is no longer supported.

Now I have said all the above, I guess you might tell me V7 won't be able to upgrade to V8, in which case my comment is a non-issue :)

**Comment by Shazwazza** [comment]
This has no affect on how we build our Nuget packages, that is a separate process, but maybe we may end up using Nuget 3, we haven't got there yet.

v8 will not be seamlessly upgradable, there are tons of breaking changes, all legacy code is being removed, most assemblies will be gone, breaking library requirements, etc... the database will be able to be migrated but you will need to rebuild and fix up your code to run v8.

File: src/Umbraco.Core/Configuration/UmbracoVersion.cs [hunk 1]
```diff
@@ -6,7 +6,7 @@ namespace Umbraco.Core.Configuration
 {
     public class UmbracoVersion
     {
-        private static readonly Version Version = new Version("8.0.0");
+        private static readonly Version Version = new Version("Release");
```
Comment (nul800sebastiaan) [inline src/Umbraco.Core/Configuration/UmbracoVersion.cs]: Seems like a mistake, this will never work.

**Comment by nul800sebastiaan** [comment]
Continued in https://github.com/umbraco/Umbraco-CMS/pull/1296

**Pull request closed without merging**